File: frontend/src/components/benchmark/content/resource_allocation/implementations/resource_allocation.java
import java.util.*;

public class ResourceAllocation {
    private Map<Integer, Set<Integer>> adjacency = new HashMap<>();
    private Set<Integer> nodes = new HashSet<>();

    public void addEdge(int u, int v) {
        adjacency.putIfAbsent(u, new HashSet<>());
        adjacency.putIfAbsent(v, new HashSet<>());
        adjacency.get(u).add(v);
        adjacency.get(v).add(u);
        nodes.add(u);
        nodes.add(v);
    }

    public void buildFromEdges(List<int[]> edges) {
        for (int[] edge : edges) {
            addEdge(edge[0], edge[1]);
        }
    }

    public double resourceAllocationScore(int u, int v) {
        if (!adjacency.containsKey(u) || !adjacency.containsKey(v)) {
            return 0.0;
        }

        double score = 0.0;
        for (int node : adjacency.get(u)) {
            if (adjacency.get(v).contains(node)) {
                int degree = adjacency.get(node).size();
                if (degree > 0) {
                    score += 1.0 / degree;
                }
            }
        }
        return score;
    }

    public List<Map.Entry<int[], Double>> scoreAllPairs() {
        List<Map.Entry<int[], Double>> scores = new ArrayList<>();
        List<Integer> nodeList = new ArrayList<>(nodes);

        for (int i = 0; i < nodeList.size(); i++) {
            for (int j = i + 1; j < nodeList.size(); j++) {
                int u = nodeList.get(i);
                int v = nodeList.get(j);
                double score = resourceAllocationScore(u, v);
                if (score > 0) {
                    scores.add(new AbstractMap.SimpleEntry<>(new int[]{u, v}, score));
                }
            }
        }

        scores.sort((a, b) -> b.getValue().compareTo(a.getValue()));
        return scores;
    }

    public List<Map.Entry<int[], Double>> topKPredictions(int k) {
        List<Map.Entry<int[], Double>> scores = scoreAllPairs();
        return scores.subList(0, Math.min(k, scores.size()));
    }
}

File: frontend/src/components/benchmark/content/resource_allocation/implementations/resource_allocation.cs
using System;
using System.Collections.Generic;
using System.Linq;

public class ResourceAllocation
{
    private Dictionary<int, HashSet<int>> adjacency = new Dictionary<int, HashSet<int>>();
    private HashSet<int> nodes = new HashSet<int>();

    public void AddEdge(int u, int v)
    {
        if (!adjacency.ContainsKey(u))
            adjacency[u] = new HashSet<int>();
        if (!adjacency.ContainsKey(v))
            adjacency[v] = new HashSet<int>();

        adjacency[u].Add(v);
        adjacency[v].Add(u);
        nodes.Add(u);
        nodes.Add(v);
    }

    public void BuildFromEdges(List<(int, int)> edges)
    {
        foreach (var (u, v) in edges)
        {
            AddEdge(u, v);
        }
    }

    public double ResourceAllocationScore(int u, int v)
    {
        if (!adjacency.ContainsKey(u) || !adjacency.ContainsKey(v))
            return 0.0;

        double score = 0.0;
        foreach (int node in adjacency[u])
        {
            if (adjacency[v].Contains(node))
            {
                int degree = adjacency[node].Count;
                if (degree > 0)
                {
                    score += 1.0 / degree;
                }
            }
        }
        return score;
    }

    public List<((int, int), double)> ScoreAllPairs()
    {
        var scores = new List<((int, int), double)>();
        var nodeList = nodes.ToList();

        for (int i = 0; i < nodeList.Count; i++)
        {
            for (int j = i + 1; j < nodeList.Count; j++)
            {
                int u = nodeList[i];
                int v = nodeList[j];
                double score = ResourceAllocationScore(u, v);
                if (score > 0)
                {
                    scores.Add(((u, v), score));
                }
            }
        }

        return scores.OrderByDescending(x => x.Item2).ToList();
    }

    public List<((int, int), double)> TopKPredictions(int k = 10)
    {
        return ScoreAllPairs().Take(k).ToList();
    }
}